lib/logstash/plugin_mixins/ecs_compatibility_support/target_check.rb

```diff
@@ -30,13 +30,45 @@ def self.included(base)
 
         ##
         # Check whether `target` is specified.
+        # For the majority of plugins, they don't have `codec`, hence does a simple check on `target`
+        # The rest of plugins, mostly input plugins, if the `codec` support `target`, check whether `codec.target` is specified
         #
         # @return [nil] if target is unspecified and ECS compatibility is disabled
-        # @return [true, false]
+        # @return [false, log_msg] when target is not set to the correct field
+        # @return [true] when target is set
```

Comment:
This change to the signature is a breaking change to `TargetCheck#target_set?`, which is provided to plugins that mix in this module. As a support dependency, it is vital that we do not introduce breaking changes, so we are going to have to come up with a non-breaking path forward.

I think I have one, but I haven't run it through its paces yet. We can add a new `TargetCheck#codec_target_set?` and change our `RegisterDecorator`'s implementation to determine which message to log (and at what level):

~~~ diff
diff --git a/lib/logstash/plugin_mixins/ecs_compatibility_support/target_check.rb b/lib/logstash/plugin_mixins/ecs_compatibility_support/target_check.rb
index ea9c744..36cacd9 100644
--- a/lib/logstash/plugin_mixins/ecs_compatibility_support/target_check.rb
+++ b/lib/logstash/plugin_mixins/ecs_compatibility_support/target_check.rb
@@ -26,6 +26,12 @@ module LogStash
             "It is recommended to set the `target` option to avoid potential schema conflicts (if your data is ECS compliant " +
             "or non-conflicting, feel free to ignore this message)").freeze
 
+        MULTIPLE_TARGETS_MESSAGE = ("This plugin and its codec both specify a `target` option, which can lead to surprising results. " +
+            "In general, it is recomended to use ONLY the codec's `target`.").freeze
+
+        PREFER_CODEC_TARGET_MESSAGE = ("Both this plugin and its codec provide a `target` option, but the codec's `target` was left unspecified. "+
+            "In general, it is recommended to only set the codec's `target`.").freeze
+
         private
 
         ##
@@ -39,6 +45,18 @@ module LogStash
           false # target isn't set
         end
 
+        ##
+        # Check whether a codec is present and specifies a target.
+        #
+        # @return [nil] if codec not present or codec does not support target
+        # @return [true,false]
+        def codec_target_set?
+          return nil unless respond_to?(:codec)
+          return nil unless codec.respond_to?(:target)
+
+          !codec.target.nil?
+        end
+
         module RegisterDecorator
 
           ##
@@ -46,7 +64,18 @@ module LogStash
           # @override
           def register
             super.tap do
-              logger.info(TARGET_NOT_SET_MESSAGE) if target_set? == false
+              if target_set? && codec_target_set?
+                # both is not good. prefer codec.
+                logger.warn(MULTIPLE_TARGETS_MESSAGE)
+              elsif target_set? && (codec_target_set? == false)
+                # prefer codec's target option
+                logger.info(PREFER_CODEC_TARGET_MESSAGE +
+                            "To do so, remove the `target` directive from this plugin and instead define the codec with " +
+                            "`codec => #{codec.config_name} { target => #{target.inspect} }`")
+              elsif (target_set? == false) && codec_target_set?.nil?
+                # target desired but unset, codec doesn't provide
+                logger.info(TARGET_NOT_SET_MESSAGE)
+              end
             end
           end
 ~~~

Reply:
:+1: on the pseudo-logic, first looking at the PR I had the concern of us doing too many warnings with `TargetCheck` included both in a input and codec plugin. but having the logic as presented in ^^ comment would avoid that ...